spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java

```diff
@@ -0,0 +1,348 @@
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.ByteType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.truncate(tinyint)";
+    }
+
+    @Override
+    public Byte produceResult(InternalRow input) {
+      int width = readWidth(input);
+      return input.isNullAt(1) ? null : invoke(width, input.getByte(1));
+    }
+  }
+
+  public static class TruncateSmallInt extends TruncateBase<Short> {
+    // magic method used in codegen
+    public static short invoke(int width, short value) {
+      return TruncateUtil.truncateShort(width, value);
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.IntegerType, DataTypes.ShortType};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.ShortType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.truncate(smallint)";
+    }
+
+    @Override
+    public Short produceResult(InternalRow input) {
+      int width = readWidth(input);
+      return input.isNullAt(1) ? null : invoke(width, input.getShort(1));
+    }
+  }
+
+  public static class TruncateInt extends TruncateBase<Integer> {
+    // magic method used in codegen
+    public static int invoke(int width, int value) {
+      return TruncateUtil.truncateInt(width, value);
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.IntegerType, DataTypes.IntegerType};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.IntegerType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.truncate(int)";
+    }
+
+    @Override
+    public Integer produceResult(InternalRow input) {
+      int width = readWidth(input);
+      return input.isNullAt(1) ? null : invoke(width, input.getInt(1));
+    }
+  }
+
+  public static class TruncateBigInt extends TruncateBase<Long> {
+    // magic function for usage with codegen
+    public static long invoke(int width, long value) {
+      return TruncateUtil.truncateLong(width, value);
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.IntegerType, DataTypes.LongType};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.LongType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.truncate(bigint)";
+    }
+
+    @Override
+    public Long produceResult(InternalRow input) {
+      int width = readWidth(input);
+      return input.isNullAt(1) ? null : invoke(width, input.getLong(1));
+    }
+  }
+
+  public static class TruncateString extends TruncateBase<UTF8String> {
+    // magic function for usage with codegen
+    public static UTF8String invoke(int width, UTF8String value) {
+      if (value == null) {
+        return null;
+      }
+
+      return value.substring(0, width);
```

Comment:
The built-in `substring` function of UTF8String is almost certainly faster than converting to `ByteBuffer` and then `CharSequence` so we can use the existing `StringBuilder` based `UnicodeUtil` method for code-point based truncation, so using that.

There's an existing function in `UTF8String`, called `toTitleCaseSlow` which uses our StringBuilder based approach, so I assume that the methods which rely on the bytes themselves and are code-point aware are faster than doing all of those conversions.

Reply:
I'd be up to using UTF8String directly. Do we have enough tests to verify the bevahior of the Spark function will match the behavior of Iceberg transform?

Reply:
This is a good point. I added a comment here to follow up and add all of our unicode truncation tests: https://github.com/apache/iceberg/pull/5431/files#r943717037

Reply:
I agree with that.

But I took a look and I actually couldn't find any test cases for `UnicodeUtil.truncateString` directly.

I looked up the tests for `TestTruncate`, but `TestTruncate#testTruncateString` has _very few_ tests - all of which are covered here.

https://github.com/apache/iceberg/blob/5a15efc070ab59eeda6343998aa065c0c9892c5c/api/src/test/java/org/apache/iceberg/transforms/TestTruncate.java#L70-L77

I was able to pull more test cases from `TestMetricsTruncation`, and we now cover 2-byte, 3-byte, and 4-byte unicode character truncation (as well as mixed byte alignment).

So I'd say we definitely have enough tests for string truncation now (at least compared to elsewhere).

Reply:
If we want to create a followup task to add more of these tests _back_ to `TestTruncate` or to a `UnicodeUtil` test suite, let me know as I could easily handle that.

At present, this code seems well tested though for using the built-in `UTF8String` tests.